plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/FacetUtil.java

```diff
@@ -0,0 +1,211 @@
+
+import com.google.cloud.tools.eclipse.util.status.StatusUtil;
+import com.google.common.annotations.VisibleForTesting;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import org.eclipse.core.resources.IContainer;
+import org.eclipse.core.resources.IFolder;
+import org.eclipse.core.resources.IProject;
+import org.eclipse.core.resources.IResource;
+import org.eclipse.core.resources.IResourceVisitor;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IPath;
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.core.runtime.OperationCanceledException;
+import org.eclipse.core.runtime.Path;
+import org.eclipse.core.runtime.SubMonitor;
+import org.eclipse.core.runtime.jobs.Job;
+import org.eclipse.jst.common.project.facet.core.JavaFacet;
+import org.eclipse.jst.common.project.facet.core.JavaFacetInstallConfig;
+import org.eclipse.jst.j2ee.project.facet.IJ2EEFacetInstallDataModelProperties;
+import org.eclipse.jst.j2ee.project.facet.IJ2EEModuleFacetInstallDataModelProperties;
+import org.eclipse.jst.j2ee.web.project.facet.IWebFacetInstallDataModelProperties;
+import org.eclipse.jst.j2ee.web.project.facet.WebFacetInstallDataModelProvider;
+import org.eclipse.jst.j2ee.web.project.facet.WebFacetUtils;
+import org.eclipse.wst.common.componentcore.internal.builder.DependencyGraphImpl;
+import org.eclipse.wst.common.componentcore.internal.builder.IDependencyGraph;
+import org.eclipse.wst.common.frameworks.datamodel.DataModelFactory;
+import org.eclipse.wst.common.frameworks.datamodel.IDataModel;
+import org.eclipse.wst.common.project.facet.core.IFacetedProject;
+import org.eclipse.wst.common.project.facet.core.IProjectFacetVersion;
+
+/**
+ * Utility class for processing facets.
+ */
+@SuppressWarnings("restriction") // For IDependencyGraph
+public class FacetUtil {
+  public static final Logger logger = Logger.getLogger(FacetUtil.class.getName());
+  
+
+  /**
+   * Adds an install action for the {@code javaFacet} to the {@code facetInstallSet} if the
+   * {@code javaFacet} does not already exist in {@code facetedProject}.
+   *
+   * @return true is the facet action was added to {@code facetInstallSet} and false if {@code javaFacet}
+   *   already exists in {@code facetedProject}
+   * @throws CoreException if {@code javaFacet} is not a Java facet
+   */
+ public static boolean addJavaFacetToBatch(IProjectFacetVersion javaFacet, IFacetedProject facetedProject,
+     Set<IFacetedProject.Action> facetInstallSet) throws CoreException {
+   if (!JavaFacet.FACET.getId().equals(javaFacet.getProjectFacet().getId())) {
+     throw new CoreException(StatusUtil.error(FacetUtil.class, javaFacet.toString() + " is not a Java facet"));
+   }
+   
+   if (facetedProject.hasProjectFacet(javaFacet)) {
+     return false;
+   }
+
+   // TODO use "src/main/java" for only maven projects
```

Comment:
This comment should appear about 4 lines below, and a link to the corresponding issue would be useful: https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/490